### PR TITLE
Proof-of-concept for the new decryption API in jwx#406

### PR DESCRIPTION
**Pull request by lack**
This switches decryption to use the new decryption PostParser hook and
avoids the deprecated msg.Decrypt() introduced in
https://github.com/lestrrat-go/jwx/pull/406

### CODE REVIEW

File: clevis.go [hunk 1]
```diff
@@ -101,23 +98,38 @@ func (p Pin) ToConfig() (Config, error) {
 
 // Decrypt decrypts a clevis bound message. The message format can be either compact or JSON.
 func Decrypt(data []byte) ([]byte, error) {
-	msg, p, err := Parse(data)
+	return jwe.Decrypt(data, "", nil, jwe.WithPostParser(clevisParser{}))
+}
+
+type clevisParser struct{}
```
Comment (anatol) [inline clevis.go:104]: Personally I am not a fan of using empty structs with a single method. Having a regular function and passing it to `jwe.WithPostParser()` would be cleaner as for me. Though upstream thinks other way unfortunately, so let's go with the  upstream way (like you did now).

File: clevis.go [hunk 2]
```diff
@@ -101,23 +98,38 @@ func (p Pin) ToConfig() (Config, error) {
...
+	p, err := pinFromMsg(ctx.Message())
 	if err != nil {
-		return nil, err
+		return err
 	}
 
 	switch p.Pin {
 	case "tang":
-		return p.Tang.decrypt(msg)
+		err = p.Tang.prepareDecryptionCtx(ctx)
```
Comment (anatol) [inline clevis.go:115]: The function `prepareDecryptionCtx()` computes the key and sets to the decryptcontext. I prefer to split these 2 operations. `prepareDecryptionCtx` receives `jwx.Message` and returns `algorithm, key, error`. And verifying/setting algo/key is done after the switch statement.

`prepareDecryptionCtx` could be renamed to something like `computeKey`.

It also makes sense to move the whole `switch` statement to a separate function under Pin struct, e.g. `Pin.computeKey()`.

Comment (anatol) [reply clevis.go:115]: I also work on making `tang` key computation more consistent with the rest of the code. With this commits https://github.com/anatol/clevis.go/tree/drop-frankenstein-ecmr all pin implementations for the `computeKey()` will return `jwa.DIRECT, []byte` where the second parameter is the key. In this case `computeKey()` signature could be further simplified to `[]byte,error`.

**Review by anatol — COMMENTED** [review]
Yeah, this use of PostParser is nicer way to deal with the struct parsing.

**Comment by anatol** [comment]
Also following test is failing:

```
--- FAIL: TestParseFailure (0.00s)
    clevis_test.go:80: 
        	Error Trace:	clevis_test.go:80
        	Error:      	Expected value not to be nil.
        	Test:       	TestParseFailure
```

File: clevis.go (hunk 1 above)
Comment (lack) [reply clevis.go:104]: Indeed, we're at their mercy :).  I would have also preferred passing a function.

Comment (lack) [reply clevis.go:104]: Perhaps we can let them know and see if they'd be willing to make a change?

File: clevis.go (hunk 2 above)
Comment (lack) [reply clevis.go:115]: I was thinking along the same lines.  I like that approach for the real implementation... I was just trying to get something to work workout putting a lot of thought into it :)

But yes, if each specific pin implemented a `recoverKey()  ([]byte, error)` that would be much cleaner!

**Comment by lack** [comment]
> Also following test is failing:
> 
> ```
> --- FAIL: TestParseFailure (0.00s)
>     clevis_test.go:80: 
>         	Error Trace:	clevis_test.go:80
>         	Error:      	Expected value not to be nil.
>         	Test:       	TestParseFailure
> ```

Oh yeah :)

I'll look into that.  

But my real-life tests, decrypting real jwe objects that were encrypted by latchset/clevis, all passed (sss, tang, and tpm2)

File: clevis.go (hunk 2 above)
Comment (anatol) [reply clevis.go:115]: In this case I suggest to go with the tang refactoring https://github.com/anatol/clevis.go/pull/9 first so your code can be simplified a bit.

Comment (lack) [reply clevis.go:115]: Agreed! I'll gladly clean this up once that's in.

File: clevis.go [hunk 3]
```diff
@@ -101,23 +99,38 @@ func (p Pin) ToConfig() (Config, error) {
...
+// Do implements a jwe.PostParser interface which must use the excracted pin config to set up the decryprion algorithm and key
+func (clevisParser) Do(ctx jwe.DecryptCtx) error {
+	p, err := pinFromMsg(ctx.Message())
 	if err != nil {
-		return nil, err
+		return err
 	}
 
+	var key []byte
 	switch p.Pin {
```
Comment (anatol) [inline clevis.go:115]: move this `switch` to a separate function `Pin.recoverKey(message)` maybe?

**Review by anatol — COMMENTED** [review]
The change looks good.

If you are ok with the new `jwx` interface we need to know lestrrat know it and ask to push his changes to `master` branch. After it is done we'll update jwx dependency to `master` and move forward with this PR.

File: clevis.go (hunk 1 above)
Comment (lack) [reply clevis.go]: They're way ahead of us: https://github.com/lestrrat-go/jwx/issues/402#issuecomment-877476832

Comment (lestrrat) [reply clevis.go]: @anatol So yeah, I agree with not wanting to use an empty struct. I _don't_ agree with explicit `func()` based parameters. Please use the `interface{}` based approach that I've put in the PR if you really must use a `func()`

**Comment by lestrrat** [comment]
So, just a warning before we all go ahead with these changes. I very well may change struct names, function names, and what not before making a release. I don't plan to yet, but please be warned and make sure you keep your go.mod pinned to a commit until I make a full release.

Just thought I might add this here since the conversation in the other thread reminded me that I wasn't 100% happy with the all the names in the new API.

**Comment by lack** [comment]
@lestrrat Makes sense :)

The good news is that this approach and proposed API works perfectly, and was easy to adapt to, whatever it ends up being called.

File: clevis.go (hunk 3 above)
Comment (anatol) [reply clevis.go:123]: I was thinking about something like this https://github.com/anatol/clevis.go/commit/7bd0e7bb7c46223088dbf3239d4492363c8fcd66

If you are ok with it ^^ then feel free to squash it into your change and push it to the PR.

**Review by anatol — APPROVED** [review]

**Comment by anatol** [comment]
Thank you for your work Jim!

**Pull request merged**